Refactor CI

frenck opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

~~WIP/Experimental refactoring of our CI to fix recent CI issues.~~

☝️ Turned out not to help with any of these. However, this refactor is still one to consider.

- It removes a pre-pare job (keeping with the same result)
- No longer relies on our "old" Azure CI containers (we can archive github.com/home-assistant/ci-azure)
- Now uses standard setup-python GitHub action task
- Splits out pre-commit, which makes those run sooner/quicker.
- Small simplifications
- Fixed a couple of tests

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

MartinHjelmare commented on .github/workflows/ci.yaml:
```diff
@@ -246,7 +263,7 @@ jobs:
           path: venv
           key: ${{ runner.os }}-${{ steps.python.outputs.python-version }}-${{
-            needs.prepare-base.outputs.python-key }}
+            needs.info.outputs.python_cache_ke }}
```
```suggestion
            needs.info.outputs.python_cache_key }}
```

cdce8p commented on .github/workflows/ci.yaml:
```diff
@@ -178,59 +186,63 @@ jobs:
           path: ${{ env.PIP_CACHE }}
           key: >-
             ${{ runner.os }}-${{ steps.python.outputs.python-version }}-${{
-            steps.generate-pip-key.outputs.key }}
+            needs.info.outputs.python_cache_key }}
```
This should probably use the pip wheel cache key -> `steps.generate-pip-key.outputs.key`

cdce8p commented on .github/workflows/ci.yaml:
```diff
@@ -178,59 +186,63 @@ jobs:
-            steps.generate-pip-key.outputs.key }}
+            needs.info.outputs.python_cache_key }}
           restore-keys: |
             ${{ runner.os }}-${{ steps.python.outputs.python-version }}-base-pip-${{ env.PIP_CACHE_VERSION }}-${{ env.HA_SHORT_VERSION }}-
```
```suggestion
            ${{ runner.os }}-${{ steps.python.outputs.python-version }}-pip-${{ env.PIP_CACHE_VERSION }}-${{ env.HA_SHORT_VERSION }}-
```
You did remove `base` from the `generate-pip-key` step output above.

cdce8p commented on .github/workflows/ci.yaml:
```diff
@@ -257,513 +269,532 @@ jobs:
-          shopt -s globstar
-          pre-commit run --hook-stage manual black --files {homeassistant,tests}/components/${{ needs.changes.outputs.integrations_glob }}/**/* --show-diff-on-failure
-
-  lint-flake8:
-    name: Check flake8
-    runs-on: ubuntu-latest
-    needs:
-      - changes
-      - prepare-base
-    steps:
```
Just a tip: If I want to disable jobs for testing, I tend to add `if: false` instead of commenting all of it. That helps reduce the diff and make it more clear where the actual changes are.

frenck replied:
> Just a tip: If I want to disable jobs for testing, I tend to add if: false instead of commenting all of it. That helps reduce the diff and make it more clear where the actual changes are.

This is a draft and work in process, toggling comments in a code editor is easier than making small expressions hidden in existing code. The comments will be gone when this PR is ready for review; so it really doesn't matter at this point.

frenck commented on tests/components/homekit/test_type_cameras.py:
```diff
@@ -501,7 +501,6 @@ async def test_camera_stream_source_configured_and_copy_codec(hass, run_driver,
     ):
         await _async_start_streaming(hass, acc)
         await _async_reconfigure_stream(hass, acc, session_info, {})
-        await _async_stop_stream(hass, acc, session_info)
```
Removed this line, as there is no stream to stop... (it never starts, as ffmpeg fails, camera doesn't exists).

cdce8p commented on .github/workflows/ci.yaml:
```diff
@@ -144,32 +158,26 @@ jobs:
 
   # Separate job to pre-populate the base dependency cache
   # This prevent upcoming jobs to do the same individually
-  prepare-base:
-    name: Prepare base dependencies
```
Tbh I did appreciate having two separate `prepare` tasks with `prepare-base` and `prepare-tests`. Usually `prepare-tests` does take quite a bit longer, so all jobs which didn't depend on it could start early and be done when pytest is run. It's just a nice experience to get most linter results pretty fast. Especially when requirements change, `prepare-base` still got a cache hit.

frenck replied:
It's an extra job that takes up time and queuing. When changing deps, it will still run the full suite anyways. That minute won't make a difference.

frenck replied:
I think it might maybe make more sense to split out the pre-commit cache creation, which will allow setting of most linters almost instantly (even faster compared to the previous situation).

bdraco reviewed (dismissed):
python changes all look good.

CI changes look sane to me, but I'm not an expert on github CI

frenck commented:
Marking it draft, going to try and see if I can improve on kickstarting the linters sooner.

frenck committed:
Restarting CI refactor

frenck committed:
Tweak

frenck committed:
More linters

frenck committed:
Fix dependency

frenck committed:
Refactor jobs on full base environment

frenck committed:
Pylint ignore missing annotations

frenck committed:
Run pytest multicore in partial runs

frenck commented:
Alright, I've soft reset all the commits, rebased, and started re-working from there again.

In the second round, I've added back a base step for just pre-commit, so those jobs will be run quickly and swiftly.

bachya reviewed (approved):
LGTM 👍🏻

frenck commented:
Thanks, @bachya 🎉
